File: 121. Best Time to Buy and Sell Stock.c
int maxProfit(int* prices, int pricesSize) {
    int min=prices[0];
    int result=0;
    int profit=0;
    
    for(int i=0;i<pricesSize;i++){
        if(prices[i]-min>profit) profit = prices[i]-min;
        if(prices[i]<min) min = prices[i];     
    }
    
    return profit;
}

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if(prices.empty()) return 0;
        int min=prices[0];
        int result=0;
        
        for(int i=1;i<prices.size();i++){
            if(prices[i]-min > result) result = prices[i]-min;
            if(prices[i]< min) min = prices[i];
        }
        return result;
    }
};

File: Algorithms/Dynamic programming/121. Best Time to Buy and Sell Stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if len(prices) <= 1:
            return 0

        l = len(prices)
        res = 0
        min_price = float('inf')

        for i in range(0, l):
            min_price = min(prices[i], min_price)
            res = max(prices[i]-min_price, res)
        return res


#         if len(prices) == 1:
#             return 0

#         maxprice = [0]*len(prices)
#         max_result = 0

#         for i in range(1, len(prices)):
#             maxprice[i] = max(maxprice[i-1] + prices[i] - prices[i-1], 0)
#             max_result = max(max_result, maxprice[i])
#         return max_result
            
